bfvmm/include/hve/arch/intel_x64/ept/memory_map.h

```diff
@@ -0,0 +1,172 @@
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+// *INDENT-OFF*
+
+namespace eapis
+{
+namespace intel_x64
+{
+namespace ept
+{
+
+class EXPORT_EAPIS_HVE memory_map
+{
+public:
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    memory_map();
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    ~memory_map();
+
+    /// Map guest physical address to host physical address with the given
+    /// page size
+    ///
+    /// @expects hpa aligned to size
+    /// @ensures
+    ///
+    /// @return Returns the leaf extended page table entry that maps gpa->hpa
+    ///
+    epte_t & map(gpa_t gpa, hpa_t hpa, uint64_t size) const;
+
+    /// Unmap a page by guest physical address
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    void unmap(gpa_t gpa) const;
+
+    /// Guest physical address to leaf extended page table entry
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the leaf extended page table entry that maps gpa->hpa
+    ///
+    epte_t & gpa_to_epte(gpa_t gpa) const;
+
+    /// Guest physical address to host physical address
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the host physical address that gpa is mapped to
+    ///
+    hpa_t gpa_to_hpa(gpa_t gpa) const;
+
+    /// Convert this memory maps page tables to a flat memory descriptor list.
+    /// NOTE: The returned memory descriptor list does not describe memory
+    /// mapped by the page tables, but rather the memory used to hold the
+    /// extended page tables themselves.
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return List of memory descriptors that describe the extended page
+    /// tables
+    ///
+    std::vector<memory_descriptor> to_mdl() const;
+
+    /// Return the base host physical address of this memory map
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return This memory maps base host physical address
+    ///
+    hpa_t hpa() const;
+
+#ifndef ENABLE_BUILD_TEST
+private:
+#endif
+    hva_t m_pml4_hva{0};
+    hpa_t m_pml4_hpa{0};
+    mutable std::mutex m_mutex;
+
+    hpa_t allocate_page_table() const;
+    void allocate_page_table(epte_t & entry) const;
+    void free_page_table(epte_t & entry) const;
+    void map_entry_to_page_frame(epte_t & entry, hpa_t hpa) const;
+
+    epte_t & gpa_to_pml4e(gpa_t gpa) const;
+    epte_t & gpa_to_pdpte(gpa_t gpa, epte_t &pml4e) const;
+    epte_t & gpa_to_pde(gpa_t gpa, epte_t &pdpte) const;
+    epte_t & gpa_to_pte(gpa_t gpa, epte_t &pde) const;
+
+    gpa_t hpa_to_gpa(hpa_t hpa, epte_t * pt, std::vector<uint64_t> pt_offsets) const;
+
+    epte_t & map_pdpte_to_page(gpa_t gpa, hpa_t hpa) const;
+    epte_t & map_pde_to_page(gpa_t gpa, hpa_t hpa) const;
+    epte_t & map_pte_to_page(gpa_t gpa, hpa_t hpa) const;
+
+    std::vector<memory_descriptor>
+    to_mdl(std::vector<memory_descriptor> mdl, epte_t * page_table) const;
+
+public:
```

Comment:
space after public, space between move and copy, and you are missing your condition flags for doxygen. In general, please make sure code like this mimics the rest of the code in the repo.